gdal/doc/source/development/rfc/rfc82_memory_recommendations.rst

```diff
@@ -0,0 +1,85 @@
+
+Summary
+-------
+
+Move the standard GDAL memory allocation strategy to RAII and allow the language handing of
+memory allocation errors to prevail.
+
+Motivation
+----------
+
+The current model of GDAL memory handling is varied, subject to error and tedious.
+Since GDAL is a C++ library, it is reasonable to take advantage of modern C++ memory
+handling practices. Encouraging a more standard model will simplify code, improve consistency,
+make it easier for new developers to contribute and reduce the likelihood of bugs.
+
+Details
+-------
+
+When GDAL came into being the C++ standard library either didn't exist or wasn't well-supported
+by many vendors. In addition, compilers were less sophisticated and tools to deal with
+memory issues were costly and difficult to use.
+
+Today, all relevant C++ compilers support the C++ standard through C++11 fully. Most support
+more recent standards as well. This allows the standard libraries to be used where
+previously C-style memory allocation was the norm. This allows GDAL code to be written with
+using a RAII model rather than a malloc/free (new/delete) model in most cases. Because RAII
+eliminates the need to free memory, early returns are simple and safe and memory leaks are
+harder to create. Exceptions can also be thrown without worrying about creating memory leaks.
+
+The existing CPL/VSI utility functions provide support for catching memory allocation errors
+and tracking programmer errors WRT memory, but this functionality has become virtually
+obsolete because
+
+    - The C++ standard library throws an exception (std::bad_alloc) when a memory error occurs,
+      allowing an application or API user to catch the (unrecoverable) error and do what
+      they wish. This eliminates any need to catch such failures locally. This exception could
+      be added to current GDAL memory allocation utilities to provide consistency for
+      existing code.
+    - The GDAL memory utilities that assist in debugging errors has largely been supplanted
+      by tools like valgrind and compiler sanitizers.
+    - Proper use of RAII eliminates the majority of memory allocation errors.
+    - Computers typically have much larger memory/VM available that makes the likelihood
+      of allocation failures rare events.  Some systems will auto-allocate VM from
+      available disk.
+
+This RFC doesn't institute any particular change.  Instead it provides recommendations for
+acceptable memory handling for changes to existing code and new additions.
+The point is to specify how memory management
+and handling in GDAL should be done if it were to be have been done from scratch in 2021.
+
+Recommendations
+---------------
+
+    - Use the C++ standard library where possible for buffer creation and general
+      data handling.
+    - Where practical, use self-created RAII types where the C++ standard library is
+      insufficient.
+    - Allow memory errors to be thrown as std::bad_alloc exceptions.
```

Comment:
This would need to be more precisely defined. Currently users (either internal to GDAL, or external) of the GDAL C++ API will never receive a C++ exception, consequently their code might not be ready for it. In some coding styles, C++ exceptions are banned (https://google.github.io/styleguide/cppguide.html#Exceptions)
Obviously all GDAL C API should catch any C++ exceptions.
Basically allowing C++ exceptions requires a careful inspection of all code paths to where it can be propagated to make sure they use themselves RAII to avoid memory leaks or other type of unfreed resources (file pointers)

Reply:
I would normally agree, but this is for all practical purposes a fatal exception WRT any GDAL processing, which isn't much different from what can happen now, which is that `abort()` is called.

Reply:
> which isn't much different from what can happen now, which is that `abort()` is called.

Memory allocation failures in GDAL, at least the larger ones normally done with VSIMalloc()/VSICalloc() (not talking about the small ones done by CPLMalloc()/CPLCalloc()) fail gracefully.

Reply:
I'm sorry but I don't understand. What does that mean to "fail gracefully"?

Reply:
> What does that mean to "fail gracefully"?

VSIxxxxx() functions related to memory allocations return a nullptr in case of failure, and the caller can exit properly, doing necessary cleanups. Like in https://github.com/OSGeo/gdal/blob/da5ad8ed4020daac5cd784ab27c5acaa899e3d90/gdal/alg/gdalwarpkernel.cpp#L5832